### ISSUE TITLE
Api to send mail, with params

### ISSUE THREAD

**Issue by vjeantet** [knadh/listmonk#42]
Hello,

  Is there a way to send, via api, an mail to a list, with a predefined template and given params ?

  I have a template with multiple placeholder {{message}} {{ticketLink}} {{title}} {{ incident_id }} and i wish to automate email sending in a external script which will consume the api with theses params :
* listNameOrID
* templateNameOrID
* params
   * title = “incident on dhcp”
   * message = “there is an incident”
   * linkTicket = “https://whatever.com/INC-5432”
   * incident_id = 354543

Can i achieve this with the last alpha ?

**Comment by knadh** [comment]
Yes, it's possible to create a campaign programmatically. https://github.com/knadh/listmonk/blob/ccd354df3f85259a0b88dffe4a51586aab01135b/handlers.go#L78

The JSON input payload would be the struct fields here: https://github.com/knadh/listmonk/blob/ccd354df3f85259a0b88dffe4a51586aab01135b/campaigns.go#L22

Templates can also be managed via APIs: https://github.com/knadh/listmonk/blob/ccd354df3f85259a0b88dffe4a51586aab01135b/handlers.go#L91

Params in the templates come from the stored subscribe metadata (arbitrary JSON). If your params are the same for the campaign and not per-subscriber, you may just want to create a template with the values pre-inserted and then start a campaign for that listID and templateID.

PS: Need to really document these in the docs!

**Comment by vjeantet** [comment]
Thanks for this fast answer, my params are the same for the campaign and not per-subscriber.

Creating a template for each campaign is not possible, the aim goal is to prepare and maintain reusable templates and use them from any scripts. If my template have to evolve, i will have to update all scripts :(

The main goal is to have a coherent template around all campaigns (alert, incident, call out, announcements, service offering...) with the power of listmonk (mail delivery, analytics and dynamic list thanks to api)

**Comment by knadh** [comment]
Let me evaluate how feasible it is to also take an arbitrary JSON payload per campaign that can then be used in templates.

**Comment by kyleshrader** [comment]
It would be nice to have this in the API docs. Currently it looks as if listmonk just manages a list of subscriptions, since there is no api docs around the campaign api.

**Comment by originswift-sys** [comment]
Ah, this one feature would complete this wonderful software. @vjeantet did you find any workarounds for template management and programmatic mailing?

**Comment by charlesmudy** [comment]
Hello,

For some reason, I'm unable to get this to work and I keep getting `Request failed with status code 400.` 

I am passing this data via API:

`data: {
	Name: "Demo",
	Subject: "Demo Subject",
	FromEmail: "Demo <noreply@mail.demo.com>",
	Body: "<raw html code inserted here>",
	ContentType: "rawhtml",
	Tags: "",
	TemplateID: 1,
	ListIDs: [15,14],
	SendAt: ""
}`

I tried looking at this: [https://github.com/knadh/listmonk/blob/ccd354df3f85259a0b88dffe4a51586aab01135b/campaigns.go#L184](https://github.com/knadh/listmonk/blob/ccd354df3f85259a0b88dffe4a51586aab01135b/campaigns.go#L184)

Can you share sample data that should be sent?

BTW, thanks for this awesome code ... amazing work.

**Comment by knadh** [comment]
@charlesmudy  the API and the field names as you've shown do not exist. There is no API currently that allows sending a single mail to an individual user. Please refer to the [docs](https://listmonk.app/docs/apis/subscribers/) for the list of supported APIs.

**Comment by charlesmudy** [comment]
@knadh thanks for the prompt response:

I am trying to send campaign to a group of people in list by sending POST to `/api/campaigns`

Looking at this #386  ... are those the fields required? I think this should be documented, the doc only has getting list of campaigns and not POST request

**Comment by knadh** [comment]
Ah, you mean the API to start a campaign. Just realised that that's missing in the docs. Will add.

```shell
curl 'http://localhost:9000/api/campaigns/$CAMPAIGN_ID/status' -X PUT \
	-H 'Content-Type: application/json;charset=utf-8' \
	--data-raw '{"status":"running"}'
```

**Comment by charlesmudy** [comment]
Sorry, still unable to get this working @knadh 

See what I am passing to the API:

`
data: {
					type: "regular",
					name: "demo",
					subject: "demo",
					from_email: "email@mail.demo.com",
					status: "running",
					body: "<p>Hello</p>",
					content_type: "html",
					template_id: 1,
					lists: [15]
				}
`

I'm just a bit lost ...

**Comment by knadh** [comment]
You have to make two API calls.

1. `POST /campaigns` to create the campaign. Refer to the [docs](https://listmonk.app/docs/apis/campaigns/).

2. Once the campaign is created, use its ID to make change it's status to `running` with the API I shared earlier.

```shell
curl 'http://localhost:9000/api/campaigns/$CAMPAIGN_ID/status' -X PUT \
	-H 'Content-Type: application/json;charset=utf-8' \
	--data-raw '{"status":"running"}'
```

**Comment by charlesmudy** [comment]
Got it working already, thanks. This will certainly be helpful for others that it requires 2 steps.

**Comment by Jjagg** [comment]
Just commenting to include the word 'transactional' so people find this issue a little easier, since it's used to track transactional mail functionality. (I found myself searching for this issue a couple times to check for progress)

**Comment by knadh** [comment]
[Transactional mailing](https://github.com/knadh/listmonk/commit/463e92d1e1353a79a8fd8d32c75fcef5e38dee5a) is now merged and will be available in the upcoming release, [v2.2.0](https://github.com/knadh/listmonk/issues/851). Will publish the API docs soon.

Took forever, but it's finally here!
